clippy_lints/src/casts/cast_sign_loss.rs

```diff
@@ -18,44 +20,150 @@ pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, cast_op: &Expr<'_>, c
+///    even number of the vairables left: `x*x*x.abs()*x.abs()*x*x`
+/// 3. Any number of any variables, but all them are guarenteed to be positive:
+///    `x.abs()*y.abs()*z.abs()`
+/// 4. Any number of any variables, ignoreing the guarenteed ones, there are even number of same
+///    variables left: `x.abs()*y.abs()*x*x*y*y`
+/// 5. Any number that calls `pow*` with even number: `x.pow(2)`
+///
+/// TODO:
+/// 1. ~A function called `expr_is_always_positive`~
+/// 2. ~A function to peel binary op (Skip ops other than Mul, Div)~
+/// 3. ~When peeling binary op, collect each expr~
+/// 4. ~For each expr, check if they are always positive, then count them if not~
+/// 5. ~Detect above cases and call it for the day~
 fn should_lint(cx: &LateContext<'_>, cast_op: &Expr<'_>, cast_from: Ty<'_>, cast_to: Ty<'_>) -> bool {
     match (cast_from.is_integral(), cast_to.is_integral()) {
         (true, true) => {
             if !cast_from.is_signed() || cast_to.is_signed() {
                 return false;
             }
 
-            // Don't lint for positive constants.
-            let const_val = constant(cx, cx.typeck_results(), cast_op);
-            if let Some(Constant::Int(n)) = const_val
-                && let ty::Int(ity) = *cast_from.kind()
-                && sext(cx.tcx, n, ity) >= 0
-            {
+            // Don't lint if `cast_op` is known to be positive.
+            if let Sign::ZeroOrPositive = expr_sign(cx, cast_op, cast_from) {
                 return false;
             }
 
-            // Don't lint for the result of methods that always return non-negative values.
-            if let ExprKind::MethodCall(path, ..) = cast_op.kind {
-                let mut method_name = path.ident.name.as_str();
-                let allowed_methods = ["abs", "checked_abs", "rem_euclid", "checked_rem_euclid"];
-
-                if method_name == "unwrap"
-                    && let Some(arglist) = method_chain_args(cast_op, &["unwrap"])
-                    && let ExprKind::MethodCall(inner_path, ..) = &arglist[0].0.kind
-                {
-                    method_name = inner_path.ident.name.as_str();
-                }
-
-                if allowed_methods.iter().any(|&name| method_name == name) {
-                    return false;
-                }
+            let (mut uncertain_count, mut negative_count) = (0, 0);
+            // Peel off possible binary expressions, e.g. x * x * y => [x, x, y]
+            let Some(exprs) = exprs_with_selected_binop_peeled(cast_op) else {
+                // Assume cast sign lose if we cannot determine the sign of `cast_op`
+                return true;
+            };
+            for expr in exprs {
+                let ty = cx.typeck_results().expr_ty(expr);
+                match expr_sign(cx, expr, ty) {
+                    Sign::Negative => negative_count += 1,
+                    Sign::Uncertain => uncertain_count += 1,
```

Comment:
Any number of uncertain can result in sign loss, no? Because we basically have no information on how many numbers of negatives there are